capa/features/extractors/ts/extractor.py

```diff
@@ -0,0 +1,105 @@
+from capa.features.extractors.base_extractor import Feature, BBHandle, InsnHandle, FunctionHandle, FeatureExtractor
+
+
+class TreeSitterFeatureExtractor(FeatureExtractor):
+    code_sections: List[TreeSitterExtractorEngine]
+    template_namespaces: List[Tuple[Node, str]]
+    language: str
+    path: str
+
+    def __init__(self, path: str):
+        super().__init__()
+        self.path = path
+        with open(self.path, "rb") as f:
+            buf = f.read()
+
+        self.language = capa.features.extractors.script.get_language_from_ext(path)
+        if self.language == LANG_TEM:
+            (
+                self.code_sections,
+                self.template_namespaces,
+            ) = self.extract_code_from_template(buf)
+        elif self.language == LANG_HTML:
+            self.code_sections = list(self.extract_code_from_html(buf))
+        else:
+            self.code_sections = [TreeSitterExtractorEngine(self.language, buf)]
+
+    def extract_code_from_template(self, buf: bytes) -> Tuple[List[TreeSitterExtractorEngine], List[Tuple[Node, str]]]:
+        template_engine = TreeSitterTemplateEngine(buf)
+        template_namespaces = list(template_engine.get_template_namespaces())
+        code_sections = list(template_engine.get_parsed_code_sections())
+
+        additional_namespaces = set(name for _, name in template_namespaces)
+        for node, _ in template_engine.get_content_sections():
+            section_buf = template_engine.get_byte_range(node)
+            code_sections.extend(list(self.extract_code_from_html(section_buf, additional_namespaces)))
+        return code_sections, template_namespaces
+
+    def extract_code_from_html(
+        self, buf: bytes, additional_namespaces: set[str] = None
+    ) -> Iterator[TreeSitterExtractorEngine]:
+        yield from TreeSitterHTMLEngine(buf, additional_namespaces).get_parsed_code_sections()
+
+    def get_base_address(
+        self,
+    ) -> Union[AbsoluteVirtualAddress, capa.features.address._NoAddress]:
+        return NO_ADDRESS
+
+    def extract_template_namespaces(self) -> Iterator[Tuple[Feature, Address]]:
+        for node, name in self.template_namespaces:
+            if node is None:
+                yield Namespace(name), NO_ADDRESS
+            else:
+                yield Namespace(name), FileOffsetRangeAddress(node.start_byte, node.end_byte)
+
+    def extract_global_features(self) -> Iterator[Tuple[Feature, Address]]:
+        yield from capa.features.extractors.ts.global_.extract_features()
+
+    def extract_file_features(self) -> Iterator[Tuple[Feature, Address]]:
+        if self.language == LANG_TEM:
+            yield from self.extract_template_namespaces()
+        for engine in self.code_sections:
+            yield from capa.features.extractors.ts.file.extract_features(engine)
+
+    def get_pseudo_main_function(self, engine: TreeSitterExtractorEngine) -> FunctionHandle:
```

Comment:
need documentation on what "pseudo main" is

Reply:
@mike-hunhoff I am not sure I understand your question. "pseudo main" is just a way to handle the script global scope. We would like to match function-level features if they are declared as global scope statements. We therefore consider the script global scope as a self-contained function scope as if all the global-scope statements were contained in a main function.